model/cpp/register_model.h

```diff
@@ -31,107 +31,102 @@
+    volatile SizeType* reg = reinterpret_cast<volatile SizeType*>(address + (bankOffset * bank));
     *reg = value;
   }
 };
 
-template<uint32_t address, uint32_t position, uint32_t width, uint32_t bankOffset = 0>
+template<typename SizeType, SizeType address, SizeType position, SizeType width, SizeType bankOffset = 0, typename = void>
 class FieldModel
 {
 public:
-  static constexpr uint32_t Width = width;
-  static constexpr uint32_t Mask = (1U << width) - 1U;
-  static constexpr uint32_t Position = position;
+  static constexpr SizeType Width = width;
+  static constexpr SizeType Mask = (1U << width) - 1U;
+  static constexpr SizeType Position = position;
 
 public:
-  static void write(uint32_t bank, uint32_t value)
+  static void write(SizeType bank, SizeType value)
   {
-    volatile uint32_t* reg = reinterpret_cast<volatile uint32_t*>(address + (bankOffset * bank));
+    volatile SizeType* reg = reinterpret_cast<volatile SizeType*>(address + (bankOffset * bank));
     *reg = (*reg & ~(Mask << position)) | ((value & Mask) << position);
   }
 
-  static uint32_t read(uint32_t bank)
+  static SizeType read(SizeType bank)
   {
-    volatile uint32_t* reg = reinterpret_cast<volatile uint32_t*>(address + (bankOffset * bank));
+    volatile SizeType* reg = reinterpret_cast<volatile SizeType*>(address + (bankOffset * bank));
     return (*reg >> position) & Mask;
   }
-
-  static_assert(position >= 0 && position <= 31, "Field position must be between 0 and 31");
-  static_assert(width >= 1 && width <= 32, "Field width must be between 1 and 32");
 };
 
-template<uint32_t address, uint32_t position, uint32_t width>
-class FieldModel<address, position, width, 0>
+template<typename SizeType, SizeType address, SizeType position, SizeType width, SizeType bankOffset>
+class FieldModel<SizeType, address, position, width, bankOffset, std::enable_if_t<bankOffset == 0>>
 {
 public:
-  static constexpr uint32_t Width = width;
-  static constexpr uint32_t Mask = (1U << width) - 1U;
-  static constexpr uint32_t Position = position;
+  static constexpr SizeType Width = width;
+  static constexpr SizeType Mask = (1U << width) - 1U;
+  static constexpr SizeType Position = position;
 
 public:
-  static void write(uint32_t value)
+  static void write(SizeType value)
   {
-    volatile uint32_t* reg = reinterpret_cast<volatile uint32_t*>(address);
+    volatile SizeType* reg = reinterpret_cast<volatile SizeType*>(address);
     *reg = (*reg & ~(Mask << position)) | ((value & Mask) << position);
   }
 
-  static uint32_t read()
+  static SizeType read()
   {
-    volatile uint32_t* reg = reinterpret_cast<volatile uint32_t*>(address);
+    volatile SizeType* reg = reinterpret_cast<volatile SizeType*>(address);
     return (*reg >> position) & Mask;
   }
-
-  static_assert(position >= 0 && position <= 31, "Field position must be between 0 and 31");
-  static_assert(width >= 1 && width <= 32, "Field width must be between 1 and 32");
```

Comment:
You could check something like `position >= 0 && position < sizeof(SizeType) * 8`